### PR TITLE
Aff_transformation_2: Add suport for reflections

### LINKED ISSUE #1216: Instancing a CGAL::Transformation of type Reflection does not work
Hello,

The instanciation of an Reflection does not seem to work.

```
#include <QCoreApplication>

#include <CGAL/Cartesian.h>
#include <CGAL/Aff_transformation_2.h>
typedef CGAL::Cartesian<double> Kernel;
typedef CGAL::Aff_transformation_2<Kernel> Transformation;

int main(int argc, char *argv[])
{
    QCoreApplication a(argc, argv);

    Transformation rational_rotate(CGAL::ROTATION,  1, 100); //OK
    Transformation translate(CGAL::TRANSLATION, CGAL::Vector_2<Kernel>(-2, 0)); //OK

    CGAL::Point_2<Kernel> A( 1, 2);
    CGAL::Point_2<Kernel> B( 3.0, 4.0);

    CGAL::Line_2<CGAL::Cartesian<double> > AB(A,B);
    Transformation Refl(CGAL::REFLECTION, AB); //KO

    return a.exec();
}
```

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

This PR adds support for 2D Reflection from a Line_2 in the Cartesian Kernel. One was already here in the Homogeneous kernel.

## Release Management

* Affected package(s):Kernel_23, Cartesian_kernel
* Issue(s) solved (if any): fix #1216
* Feature/Small Feature (if any): https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Aff_transformation_2/Reflections
* pre-approuved by @lrineau, 2019-06-17
* Link to compiled documentation (obligatory for small feature) [link](https://cgal.geometryfactory.com/~mgimeno/Aff-transformation_2-Reflection/Kernel_23/classCGAL_1_1Aff__transformation__2.html)
* License and copyright ownership:
// Copyright (c) 2018  
// Utrecht University (The Netherlands),
// ETH Zurich (Switzerland),
// INRIA Sophia-Antipolis (France),
// Max-Planck-Institute Saarbruecken (Germany),
// and Tel-Aviv University (Israel).  All rights reserved.

### CODE REVIEW

File: Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    line(l)
+  {
+    if(l.a() == 0) 
+      t = -Vector_2(0, l.c()/l.b());
+    else
+      t = -Vector_2(l.c()/l.a(),0);
+    
+    Vector_2 l_to_v = l.to_vector(),
+        Ox = Vector_2(1,0);
+    FT scal = l_to_v * Ox; //= |L|*cos(a)
```
Comment (lrineau) [inline Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h]: As `Ox` is the vector (1,0), `l_to_v*Ox` is just `l_to_v.x()`.

File: Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+typedef typename CGAL::Line_2<R>                 Line_2;
+  typedef typename Aff_t_base::Aff_transformation_2   Aff_transformation_2;
+  typedef Aff_transformation_repC2<R>      Transformation;
+  typedef Reflection_repC2<R>              Reflection;
+  typedef Scaling_repC2<R>                 Scaling;
+  typedef Rotation_repC2<R>                Rotation;
+  typedef Translation_repC2<R>             Translation;
+
+  Reflection_repC2(const Line_2 &l) :
+    line(l)
```
Comment (lrineau) [inline Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h]: Why is the line copied, here?

File: Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+          r.cosinus_*(t13()-r.t.x()) + r.sinus_*(t23()-r.t.y())+r.t.x(),
+          
+          r.sinus_*cosinus_ - r.cosinus_*sinus_, 
+          r.sinus_*sinus_+r.cosinus_*cosinus_, 
+          r.sinus_*(t13()-r.t.x()) -r.cosinus_*(t23()-r.t.y())+r.t.y());
+  }
+  
+  Aff_transformation_2  inverse() const
+  {
+    return Aff_transformation_2(REFLECTION, line);
```
Comment (lrineau) [inline Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h]: The member `line` seems to be used only here, but the inverse of a reflection is itself. That means you can return `Aff_transformation_2(*this)`. Then there is no need, I think, to store the line in the transformation.

Comment (maxGimeno) [reply Cartesian_kernel/include/CGAL/Cartesian/Reflection_rep_2.h]: It is not possible, `Reflection_rpC2` is not an `Aff_transformation_2`, so the constructor you are talking about doesn't exist.

**Comment by lrineau** [comment]
That is small feature. Similar to [Features/Small Features/Classification: Neural Network and parallelized Random Forest][1].

[1]: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Classification:_Neural_Network_and_parallelized_Random_Forest

**Comment by maxGimeno** [comment]
Oh, thank you !

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-105.shtml.

File: Polygon/include/CGAL/Polygon_2/Polygon_2_impl.h [hunk 4]
```diff
@@ -94,7 +94,7 @@ operator>>(std::istream &is, Polygon_2<Traits_P,Container_P>& p)
 {
   int n = 0; // number of vertices
   is >> n;
-  typename Traits_P::Point_2 point;
+  typename Traits_P::Point_2 point(0,0);
```
Comment (lrineau) [inline Polygon/include/CGAL/Polygon_2/Polygon_2_impl.h]: That initialization only hides the warning, but the compiler is right: there is a problem with the code. The real problem is below:
```c++
      for (int i=0; i<n; i++) {
        is >> point;
        p.push_back(point);
      }
```
It should be:
```c++
      for (int i=0; i<n; i++) {
        if(is >> point) {
          p.push_back(point);
        } else {
          return is;
        }
      }
```

Comment (maxGimeno) [reply Polygon/include/CGAL/Polygon_2/Polygon_2_impl.h]: Ow ok.

File: Polygon/include/CGAL/Polygon_2/Polygon_2_impl.h [hunk 5]
```diff
@@ -95,15 +95,20 @@ operator>>(std::istream &is, Polygon_2<Traits_P,Container_P>& p)
...
   is >> n;
   typename Traits_P::Point_2 point;
- 
+
   if (is) {
       p.erase(p.vertices_begin(),p.vertices_end());
       for (int i=0; i<n; i++) {
-        is >> point;
-        p.push_back(point);
+	if( is >> point )
```
Comment (lrineau) [inline Polygon/include/CGAL/Polygon_2/Polygon_2_impl.h]: The indentation looks strange.

**Comment by maxGimeno** [comment]
I propose to postpone this . No one has done a review yet.

**Comment by lrineau** [comment]
This PR conflicts locally with #3939, merged recently in `master` and release branches. I will update this PR...

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-71.shtml

**Pull request merged**